javascript/ql/src/semmle/javascript/CharacterEscapes.qll

```diff
@@ -0,0 +1,97 @@
+/**
+ * Provides classes for reasoning about character escapes in literals.
+ */
+
+import javascript
+
+module CharacterEscapes {
+  /**
+   * Provides sets of characters (as strings) with specific string/regexp characteristics.
+   */
+  private module Sets {
+    string sharedEscapeChars() { result = "fnrtvux0\\" }
+
+    string regexpCharClassChars() { result = "bBcdDpPsSwW" }
+
+    string regexpBackreferenceChars() { result = "123456789k" }
+
+    string regexpMetaChars() { result = "^$*+?.()|{}[]-" }
+  }
+
+  /**
+   * Gets the `i'th` character of `raw`, which is preceeded by an uneven number of backslashes.
+   */
+  bindingset[raw]
+  string getAnEscapedCharacter(string raw, int i) {
+    result = raw.regexpFind("(?<=(^|[^\\\\])\\\\(\\\\{2}){0,10}).", _, i)
+  }
+
+  /**
+   * Holds if `n` is delimited by `delim` and contains `rawStringNode` with the raw string value `raw`.
+   */
+  private predicate hasRawStringAndQuote(
+    DataFlow::ValueNode n, string delim, ASTNode rawStringNode, string raw
+  ) {
+    rawStringNode = n.asExpr() and
+    raw = rawStringNode.(StringLiteral).getRawValue() and
+    (
+      raw.charAt(0) = "'" and delim = "'"
+      or
+      raw.charAt(0) = "\"" and delim = "\""
```

Comment:
Can't we just use `raw.charAt(0) = delim` here? `rawStringNode` is known to be a `StringLiteral`, so its first raw character has to be either `'` or `"`.